### ISSUE TITLE
Feature Request: batch (and link) create_transfers with create_accounts

### ISSUE THREAD

**Issue by wilsonianb** [coilhq/tigerbeetle#29]
I might be interested in being able to batch creating an account with creating a transfer involving the account.
I'd like for an account to not be created if the transfer fails.

**Comment by jorangreef** [comment]
Thanks @wilsonianb!

TigerBeetle was almost going to support mixed batch event types (e.g. a batch of accounts/transfers/commits) but we decided not to in the end to keep things simple and to avoid polymorphism.

Having the batch contain only the same type, means we can reduce branch mispredicts on the runtime batch event type (as we loop through a few thousand we don't need to inspect the type and switch on the type), use the same state machine instruction code that's already cached in the instruction cache for everything in the batch, and use the same result type. It also helps with assertions, e.g. we can know that the batch input size should be a multiple of the input type and that the batch output size should be a multiple of the output type etc.

If an account does not exist, then any subsequent transfer would fail. In the other direction, though, as you described, if the account is created and then a transfer fails, this would need the account to be cleaned up, or left in the system for auditing. That's might not be a problem if there are many more transfers than accounts.

**Comment by wilsonianb** [comment]
That all makes sense to me.

In my setup, an application-level account can have from zero to four optional tigerbeetle accounts to track various types of balances.
Some options for how the tigerbeetle accounts are created include:

1. create all four tigerbeetle accounts when creating application-level account
2. create each tigerbeetle account only when it is first involved in a transfer, and if the initial transfer fails:
  ~a. rollback tigerbeetle account creation by batching it with the transfer~
  b. orphan the tigerbeetle account (create a new one if a subsequent "initial" transfer is attempted)
  c. keep the tigerbeetle account (map it to the application-level account)

[I'm opting for 2c](https://github.com/coilhq/rafiki/pull/51/commits/950d6b0958261add4ad70a6d1611883ad843fd9e) which would limit the number of tigerbeetle accounts to `<= 4 * nAppAccounts`

**Comment by wilsonianb** [comment]
It's been pointed out that 2a would require pre-mapping new tigerbeetle account(s) to the application-level account before attempting the batched account creation + initial transfer. Otherwise, a new tigerbeetle account could have a successful initial transfer but fail to be mapped, causing money to lost. However, that would introduce the possibility of mapping a tigerbeetle account that is never successful created.

2c's order of operations is preferable:

1. create tigerbeetle account
2. map tigerbeetle account to application-level account
3. attempt transfer

There _is_ still a risk of a tigerbeetle account being orphaned due to a failure between steps 1 and 2, as well as the newly created and mapped account's initial transfer failing, but IMO both of those outcomes are more tolerable than 2a's alternatives.

tl;dr 2c is still the way to go.

**Comment by sentientwaffle** [comment]
> There is still a risk of a tigerbeetle account being orphaned due to a failure between steps 1 and 2, as well as the newly created and mapped account's initial transfer failing, but IMO both of those outcomes are more tolerable than 2a's alternatives.

The orphaned account doesn't hold any money, so I think that's fine.

**Comment by jorangreef** [comment]
It's also okay and good to leave the 128 byte account around (without deleting it) as part of the audit trail. Plus you have the complexity of these limited to `<= 4 * nAppAccounts` so that sounds peachy!
